sw/device/tests/sim_dv/i2c_tx_rx_test.c

```diff
@@ -0,0 +1,205 @@
+
+  // Temporary hack that connects i2c to a couple of open drain pins.
+  CHECK_DIF_OK(dif_pinmux_input_select(&pinmux,
+                                       kTopEarlgreyPinmuxPeripheralInI2c0Scl,
+                                       kTopEarlgreyPinmuxInselIob11));
+  CHECK_DIF_OK(dif_pinmux_input_select(&pinmux,
+                                       kTopEarlgreyPinmuxPeripheralInI2c0Sda,
+                                       kTopEarlgreyPinmuxInselIob12));
+  CHECK_DIF_OK(dif_pinmux_output_select(&pinmux, kTopEarlgreyPinmuxMioOutIob11,
+                                        kTopEarlgreyPinmuxOutselI2c0Scl));
+  CHECK_DIF_OK(dif_pinmux_output_select(&pinmux, kTopEarlgreyPinmuxMioOutIob12,
+                                        kTopEarlgreyPinmuxOutselI2c0Sda));
+
+  // I2C speed parameters.
+  dif_i2c_timing_config_t timing_config = {
+      .lowest_target_device_speed = kDifI2cSpeedFastPlus,
+      .clock_period_nanos = kClockPeriodNanos,
+      .sda_rise_nanos = kI2cRiseFallNanos,
+      .sda_fall_nanos = kI2cRiseFallNanos,
+      .scl_period_nanos = kI2cClockPeriodNanos};
+
+  dif_i2c_config_t config;
+  CHECK_DIF_OK(dif_i2c_compute_timing(timing_config, &config));
+  CHECK_DIF_OK(dif_i2c_configure(&i2c, config));
+  CHECK_DIF_OK(dif_i2c_host_set_enabled(&i2c, kDifToggleEnabled));
+  CHECK_DIF_OK(
+      dif_i2c_set_watermarks(&i2c, kDifI2cLevel30Byte, kDifI2cLevel4Byte));
+  CHECK_DIF_OK(
+      dif_i2c_irq_set_enabled(&i2c, kDifI2cIrqFmtWatermark, kDifToggleEnabled));
+  CHECK_DIF_OK(
+      dif_i2c_irq_set_enabled(&i2c, kDifI2cIrqRxWatermark, kDifToggleEnabled));
+  CHECK_DIF_OK(dif_i2c_irq_set_enabled(&i2c, kDifI2cIrqTransComplete,
+                                       kDifToggleEnabled));
+
+  // Randomize variables.
+  uint8_t byte_count = rand_testutils_gen32_range(1, 64);
+  uint8_t device_addr = rand_testutils_gen32_range(0, 16);
+  uint8_t expected_data[byte_count];
+  LOG_INFO("Loopback %d bytes with device %d", byte_count, device_addr);
+
+  // Controlling the randomization from C side is a bit slow, but might be
+  // easier for portability to a different setup later.
+  for (uint32_t i = 0; i < byte_count; ++i) {
+    expected_data[i] = rand_testutils_gen32_range(0, 0xff);
+  };
+
+  // Write expected data to i2c device.
+  CHECK(!fmt_irq_seen);
+  i2c_testutils_wr(&i2c, device_addr, byte_count, expected_data, false);
+
+  uint8_t tx_fifo_lvl, rx_fifo_lvl;
+
+  // Make sure all fifo entries have been drained.
+  do {
+    CHECK_DIF_OK(dif_i2c_get_fifo_levels(&i2c, &tx_fifo_lvl, &rx_fifo_lvl));
+  } while (tx_fifo_lvl > 0);
+  CHECK(fmt_irq_seen);
+  fmt_irq_seen = false;
+
+  // Read data from i2c device.
+  CHECK(!rx_irq_seen);
+  i2c_testutils_rd(&i2c, device_addr, byte_count);
+
+  // Make sure all data has been read back.
+  do {
+    CHECK_DIF_OK(dif_i2c_get_fifo_levels(&i2c, &tx_fifo_lvl, &rx_fifo_lvl));
+  } while (rx_fifo_lvl < byte_count);
+  CHECK(rx_irq_seen);
+
+  uint8_t byte;
```

Comment:
I would move this declaration inside the `for` loop.

Reply:
fixed